File: CP_programming/src/Sorting/quickSort.cpp
#include <cstdio>

/*Notes : 
Worst time complexity : O(n^2) 
Reason : when pivot is not chosen well, usually we have unbalanced partitions.
So, instead of log(n) recursive depth, we have recursive depth of n. 
And at each recursive depth we have n + n-1 + n-2 ... 1 comparison -> n^2
Average and best time complexity : O(nlogn) -> divide and conquer
Space complexity : O(1) for partition
                   O(logn) due to stack depth
                   O(n) in worst case [unbalanced partitions]
*/

void swap(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

int partition(int arr[], int low, int high)
{
     int i = low-1;
     int pivot = arr[high]; // pivot element
     for(int j=low;j<high;j++)
     {
        if (arr[j] < pivot)
        {
            i++;
            swap(arr[i], arr[j]);

        }
     }
     swap(arr[i+1],arr[high]);
     return i+1;
}

void quickSort(int arr[], int low, int high)
{
    if (low < high)
    {
        int pivot = partition(arr, low, high);
        quickSort(arr, low,pivot-1);
        quickSort(arr, pivot+1, high); // divide and conquer
    }
}

int main()
{
    int arr[] = {4,10,13,10,7,1,19,0};
    int size = sizeof(arr)/sizeof(arr[0]);
    quickSort(arr, 0,size-1);
    for(int i=0;i<size;i++)
    {
        printf("%d ",arr[i]);
    }
    printf("\n");
    return 0;
}

File: DSA_basics/scripts/quick_sort.py
# quick sort algorithm implementation in python
# quick sort is a divide and conquer algorithm which relies on recursion
# it has a average time complexity of O(nlogn)
# it is much faster than bubble , insertion and selection sort on average
# in the worst case, the time complexity is O(n^2)

# the first step is to implement partition function
# the partition function is used to partition the array into two parts
# the first part contains all the elements that are less than the pivot
# the second part contains all the elements that are greater than the pivot
# the pivot is the element that is being compared with the elements in the array

import random 
import time

def partition(arr, left_pointer, right_pointer):
    pivot_index = right_pointer # any index could be chosen as pivot
    pivot = arr[pivot_index]
    right_pointer -= 1
    while True:
        while arr[left_pointer] < pivot:
            left_pointer+=1
        while arr[right_pointer] > pivot:
            right_pointer -= 1
        if left_pointer >= right_pointer:
            break
        else:
            arr[left_pointer], arr[right_pointer] = arr[right_pointer], arr[left_pointer]
            left_pointer += 1
    arr[left_pointer], arr[pivot_index] = arr[pivot_index], arr[left_pointer]
    return left_pointer # returned for quick sort function

# once we have partition method, we have a way of diving the array into two parts
# so, keep dividing the subarrays this way till you reach base condition - when length is 0 or 1.
def quicksort(arr, left_index, right_index):
    if left_index >= right_index:
        return 
    pivot_index = partition(arr, left_index,right_index)
    quicksort(arr, left_index,pivot_index-1)
    quicksort(arr, pivot_index+1, right_index)


# test
arr = [8,4,0,3,2,9]
quicksort(arr,0,len(arr)-1)
print(arr)

# some tests to compare efficiency of quick sort and python's build in implementation
SIZE = 100000
# 1. average case
arr = [random.randint(0,1000) for _ in range(SIZE)]
# python sort time
arr1 = arr.copy()
start_time = time.time()
arr1.sort()
end_time = time.time()
print("python's sort function time : average case = ", end_time-start_time)

arr2 = arr.copy()
start = time.time()
quicksort(arr2, 0, len(arr2)-1)
end_time = time.time()
print("Quicksort function time : average case = ", end_time-start_time)

# 2. Best case
arr = [range(0,SIZE)]
arr1 = arr.copy()
start_time = time.time()
arr1.sort()
end_time = time.time()

print("python's sort function time : best case = ", end_time-start_time)

arr2 = arr.copy()
start_time = time.time()
quicksort(arr2, 0, len(arr2)-1)
end_time = time.time()
print("Quicksort function time : best case = ", end_time-start_time)


